### DISCUSSION TITLE
getting_started/first_2d_game/07.finishing-up

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#16] [User-contributed notes]
# getting_started/first_2d_game/07.finishing-up

We have now completed all the functionality for our game. Below are some remaining steps to add a bit more "juice" to improve the game experience. Feel free to expand the gameplay with your own ide...

https://docs.godotengine.org/en/latest/getting_started/first_2d_game/07.finishing-up.html

**Comment by ninlar** [comment]
It appears that the node types are now **AudioStreamPlayer2D** instead of just **AudioStreamPlayer**. I had to correct that in the code to get the music to play. Very awesome to complete this tutorial.

**Reply by Calinou** [reply]
Both AudioStreamPlayer and AudioStreamPlayer2D still exist in Godot 4. AudioStreamPlayer should be used for music playback, as it's non-positional, while AudioStreamPlayer2D should be used for positional audio (i.e. most sound effects).

**Reply by ninlar** [reply]
Whoops I see it now. Not sure how I missed that node :(

**Comment by K-Princ** [comment]
how can add touchscreen input to this if i export this game to android?

**Reply by timhutton** [reply]
One way is to move the player towards where the user is touching. In `player.gd`, add:

```gdscript
var touch_is_down = false
var touch_target

func _input(event):
	if event is InputEventScreenTouch:
		if event.pressed:
			touch_is_down = true
			touch_target = event.position
		elif event.released:
			touch_is_down = false
	elif event is InputEventScreenDrag:
		touch_target = event.position
```

and then in `_process()` you can add:

```gdscript
	if touch_is_down and position.distance_to(touch_target) > 5:
		velocity = touch_target - position
```

**Comment by sync-oz** [comment]
I made a small addition that may be suitable for here, which is hiding the player when the app runs and showing them when the game starts. I thought seeing the player on the title screen was odd.

Added 
```gdscript
func new_game():
    ...
	$Player.show()

func _ready():
	$Player.hide()
```

**Reply by ZakariGames** [reply]
This was already part of the tutorial actually! :D

Right before adding the collisions in the "Coding the Player" section is when we add the hide() function to the player's _ready() function.

> "When you're sure the movement is working correctly, add this line to _ready(), so the player will be hidden when the game starts:"

>"hide()"

And then at the very end of the "Coding the Player" section, the show() function is added to the player's start() function:

> "The last piece is to add a function we can call to reset the player when starting a new game."

> func start(pos):
	position = pos
	show()
	$CollisionShape2D.disabled = false

**Comment by PortableFox** [comment]
For the sound effects section mine was missing the Loop checkbox. However based on this [forum post](https://forum.godotengine.org/t/i-cant-enable-loop-using-audiostreamplayer2d/1860) I was able to do this by connecting the AudioStreamPlayer's finished() node to main and telling it to start playing the music when the audio finished signal fired like so

```gdscript
func _on_music_finished() -> void:
	$Music.play()
```

**Reply by Calinou** [reply]
Note that a script-based approach to audio looping may not perform gapless playback correctly, so you might hear a small amount of silence between each loop.

**Reply by islam3zzat** [reply]
have u tried enabling the looping and reimporting the resource?
This has worked for me

<img width="1678" alt="Screenshot 2024-09-15 at 3 13 09 PM" src="https://github.com/user-attachments/assets/903b9465-cc00-415c-9698-c842c78f1f52">

**Reply by associateone** [reply]
You have to make it unique as the tutorial says in order to enable the Loop checkbox.

**Reply by associateone** [reply]
All audio is automatically imported with the Loop setting disabled. If you want the music to loop seamlessly, click on the Stream file arrow, select Make Unique, then click on the Stream file and check the Loop box.

**Reply by MartinDelille** [reply]
Awesome tutorial!

Regarding the looping, I was not very happy to discover that the audio file looked to be embedded in the `main.tscn` file:

```tscn
[sub_resource type="OggPacketSequence" id="OggPacketSequence_atmdh"]
packet_data = Array[Array]([[PackedByteArray("AXZvcmJpcwAAAAACRKwAAP/////Q3QYA/////7gB")], [PackedByteArray("A3ZvcmJpcx0AAABYaXBoLk9yZyBsaW...
granule_positions = PackedInt64Array(0, 0, 1856, 5568, 9664, 13760, ...
sampling_rate = 44100.0
```

I noticed it happened when I clicked "Make unique". I managed to enable the loop playback by editing the "art/House In a Forest Loop.ogg.import" file when I changed loop to true.

**Reply by mark-wiemer** [reply]
Ugh, I've followed all these instructions and still nothing :(

1. Godot 4.3 with GDScript
1. Enable Loop in the Audio Stream Importer:
   
   ![loop enabled](https://github.com/user-attachments/assets/7ed829f4-0852-46f8-83d7-12bde3a0aa9d)

1. Reimport
1. Make Unique in AudioStreamPlayer > Stream > down arrow > Make Unique:

   ![stream made unique](https://github.com/user-attachments/assets/7155f23f-8345-4155-9fa8-fedb37e5f7b8)

At this point I expect a loop button to appear somewhere in the editor, but no luck. I've tested the game and the music ends, it doesn't loop :(

![no loop property in inspector](https://github.com/user-attachments/assets/97306e83-92aa-46ac-a96b-32d464eed7e0)

![filter for loop no results](https://github.com/user-attachments/assets/6e0e439e-cf99-4dbe-acbd-63b9d100adc8)

**Reply by mark-wiemer** [reply]
That said, looping via code works fine :) and you don't need to make it unique either, so that avoids the large embedding in the `.tscn` file mentioned earlier

**Reply by markusmeier345** [reply]
It seems that the Looping option has moved to the Parameters section

**Reply by LEDBug** [reply]
I had some trouble with this so explaining what I missed. I also found the "Looping" Parameter under the "Parameters" section but that didn't loop the audio (I haven't tried also checking 'On'). 
What did work was clicking on the audio file (the wavy graphic with the name OggVorbis) NOT the down arrow next to it. This is after "Make Unique" is selected. That then opens up a bigger panel that includes the "Loop" option.

**Reply by mbelaldev** [reply]
i also had trouble at first, but if you clicked on   'OggVoribs'       you will see new options inlucding loop checkbox

**Comment by KKaminski93** [comment]
I think it would make sense to add a little bit to the end of this section that describes how to export the game so you can share/playtest it with a friend.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/9976

**Comment by SteeveBlanc** [comment]
to export the game and show it proudly : 
Project/Export
Add... 
Windows runnable / Mac
it says some config files are missing
add them with the option on the right, Godot downloads automatically from best source
ok, re-open Project/Export
this time you are allowed to press Export project
create a subfolder and choose it
Go.
4 files are generated (on Windows), 3 are necessary but the console.exe seems optional
you can zip the folder with 3 files and send it

**Comment by thomzd** [comment]
While pressing the button I get this error:

E 0:00:01:0767   event_get_action_status: The InputMap action "start_game" doesn't exist. Did you mean "ui_text_caret_document_start"?
  <C++ Error>    Condition "!E" is true. Returning: false
  <C++ Source>   core/input/input_map.cpp:262 @ event_get_action_status()


screenshot: https://puu.sh/KfJ4y/1c74694461.png

**Reply by thomzd** [reply]
EDIT: even though I have the start_game function in my script:

signal start_game
#...
func _on_start_button_pressed() -> void:
	$StartButton.hide()
	start_game.emit()

**Reply by jsonjunkie** [reply]
seems like you might need to add the `start_game` Input Map. go to Project > Project Settings > Input Map and then add new action named `start_game`

**Comment by jessodum** [comment]
I don't know if this is something new in 4.3 (I'm completely new to Godot as of this week), but I think I found easier instructions to enable the music looping.

Instead of this:
> If you want the music to loop seamlessly, click on the Stream file arrow, select Make Unique, then click on the Stream file and check the Loop box.

... do this instead:

In the Inspector for the Music node, expand the **Parameters** under AudioStreamPlayer and make sure both the "Looping" and "On" checkboxes are checked.

**Reply by ElectricZax** [reply]
This is what I used. I didn't even find the first one and I thought it was me. But your comment made me look for it again and I found it.

Thank you.

**Reply by Aspherioth** [reply]
The option is still there. After clicking the arrow and choosing "Make Unique" you can click on the stream (to the left of the arrow) and it will expand with options for the loaded stream. Took me a while to realise this too. The UI looks exactly like the one on the screenshot from the tutorial once you click the stream.

**Comment by ElectricZax** [comment]
I don't know what is the proper protocol here... but can we have a separate page or section or maybe a Discord group where we can discuss various ways we improve and build on the game and how to do new things?

For example I added a button with function "boost", so when I press it the speed of my creep will double. 
if Input.is_action_pressed("move_right"):
		velocity.x += 1*boost

But I can't make an "if" condition to increase the speed when score > 30 for example (because I don't know how to setup the if rule to check the score).

**Reply by Kushal-Shankhapal** [reply]
The score variable is stored in the script attached to the main node (main.gd) and is updated by calling the update_score() method of hud.gd.

So you can emit a signal from main when the score goes above 30, connect that signal to your player node, where you can apply the boost or increase the speed.

**Comment by ElectricZax** [comment]
I could not make the "game_start" button to work. (I skipped part of the instructions, because I thought it was related to the controller support). 
In the end I added to the main script the below code and it worked.

```gdscript
func _process(delta):
	if Input.is_action_pressed("start_game"):
		new_game()
```

**Comment by ItsCorruptivity** [comment]
My game is not working.
The MObs are off scale and rotation I cant fix it
Also instead of running the main scene as its supposed to it runs the player scene when i press the run button

**Comment by Eilliw11** [comment]
I'm seeing a lot of little snips and pieces of the finished script we're supposed to write, but where these pieces go is pretty vague, and I'm almost certain not complete. It looks like there is an assumption that once you've got something figured out in one place, you'll automatically know what goes in somewhere else without any instruction. That's not really the case. Does the completed project exist somewhere to compare what I've done to a known good example? Then I can see if something is in the wrong place or if anything is missing. I can go through line by line, and see it's doing everything I told it to do. It just isn't doing everything the example is doing. Without a known good script to compare it to, I can't see the mistake neither I nor the debugger caught.

**Reply by Eilliw11** [reply]
Oh! Found it. Godot-demo-projects/2d/dodge_the_creeps. Found all three scripts right there. Not even a minute after asking, I found it. But if anyone is just a LITTLE off, there is a known good example. Literally one letter off for mine.

**Comment by schaefchenaufderweide** [comment]
After finishing, my player suddenly disappeaed. After this i am convinced, godot is not for me. It would really be helpful, if you shard the complete code.

**Reply by Calinou** [reply]
You can find the completed demo project here: https://godotengine.org/asset-library/asset/2712

**Reply by Huzal** [reply]
I had the same problem and it turns out I was dumb and skipped this line during the Background portion of this page:

Make it the first node under Main so that it will be drawn behind the other nodes.

Kind of important if you want to see what is happening to not make the background a foreground :P

Though I'm not sure why only the player gets drawn under the rectangle while the HUD and Mobs are still over top of it.

**Comment by Fares-Charni** [comment]
Before exporting (process is explained in other comments, documentation or this youtube video: https://www.youtube.com/watch?v=31W2ti676RE ) you should assign a main node. To do so, Go to "Project Parameters -> General -> Application -> Run" and select Main as you main node. This will solve the problem with running your executable.
On the other side, don't forget to put ColorRect as the first child of Main so it doesn't cover Player.
Otherwise, very cool tutorial , happy to discover godot :)

**Comment by Aapuji** [comment]
The art, music, sounds, etc. are all really cool for a noob like me. How would I go about creating art and sounds for games I make? What would you suggest? How did the creators of this tutorial make them? How would you suggest I learn how to make good art/music?

**Reply by LAyaJnamiJ** [reply]
With all creative skills, it all boils down to a few things:

1. What kind of art / SFX / music is suitable for your game? This tutorial follows a more relaxed vibe with easy listening and cute art. Does this kind of vibe suit the game of your vision?
2. What medium do you want to use to create them? With art, you could draw it physically or digitally; this tutorial drew them digitally. With music, you could either record yourself playing an instrument, or create music digitally. 
3. What kind of software do you want to use? Like game engines, different illustration software and digital audio workstations (DAW) offer different pros and cons. Play around with few of these and decide which one feels the best. Unlike Godot, most of those require purchasing licenses, but do offer free trials for testing the waters.
4. Go through tutorials to get some guidance and head starts, like this one you just went through. (Well, it's over 2 months at time of writing.) Get a feel of how people do things.
5. Lastly, akin to creating games, doing artistic creative stuff is never a one-day thing; go through it repeatedly and only by creating you get to learn.

Hope this helps a bit. Obviously not exhaustive as there is only so much that can be fit into a reply; feel free to reach out if anything. Thanks.

**Comment by hansintheair** [comment]
So, just to make sure I understood the shortcut part correctly, by setting the shortcut property on the StartButton of the HUD to start_game, and mapping start_game to Keyboard Spacebar or Controller button A, hitting the shortcut acts like a `pressed` event for the button?

**Comment by RobbiewOnline** [comment]
I hit a very odd issue, after setting the shortcut for the Start button to start_game, I found that the game wouldn't start when pressing the enter key, but worse was that clicking the button with the mouse only partially started the game.

The Start button would disappear, but the enemy never displayed and the music didn't play.

I quit & restarted the game, it consistently froze after hiding the Start button.

I decided to debug this by adding a breakpoint to the new_game function, but, when debugging the game worked! I then quit + restarted the game and then it would play by using the mouse by clicking on the Start button every time.

In desperation I reverted my code using GitHub, repeated the process and got exactly the same weird quirk, I had to debug to unblock the start_game function! 

As of now, the game consistently runs.

I still however don't have the shortcut setup for the ketboard, even though I followed the instructions.

I'm using GoDot 4.3 on a MacBook Air (Apple M2).  I'll try discord to ask for tips and will reply here if I get an answer.

**Reply by lukegrob** [reply]
Slightly delayed but i just had a similar situation.  I read your more recent comment also, but..

I use F6 to run the game to play it (which runs current scene) when i should probably be pressing F5 (run the main scene of the project).  Because I was in the hud window to access the startButton (to add the shortcut), i was pressing f6 and actually running the hud, which looks almost the same except no bg colour and no mobs spawn -- i press the button, the button disappear, as if the game is running, but then nothing else happens for me, because of course I am running the wrong scene.

Once I made sure i was running the main scene (or using f5), it worked every time for me.

This may not be your issue (I see you went further with trying to debug), but as my experience was similar and in the same place, I just wanted to say, just in case!

**Reply by RobbiewOnline** [reply]
Amazing - thank you for highlighting this.

Funny enough I created a comment to one of their tutorials where I've given similar advice!

It's likely other n00bs will hit it, so hopefully being prominent   in the tutorial will be an aha moment.

It feels now that the getting started user guide needs to warn n00bs that this may happen and to keep an eye out for it 😅

**Comment by tangramor** [comment]
I cannot make the shortcut part work until I change it to **InputEventKey** and configure it to use **Enter**
![image](https://github.com/user-attachments/assets/0f012c71-2c30-491a-b20e-267b618ea489)

**Reply by RobbiewOnline** [reply]
Did you experience the weird issue I had (comment prior to yours)

> I found that the game wouldn't start when pressing the enter key, **but worse was that clicking the button with the mouse only partially started the game.**

**Reply by tangramor** [reply]
no...My env is Windows 10 with GoDot 4.3

**Comment by Jacinto73** [comment]
I finished it. Everything works. I have tried so many times on previous versions to get this going and there was always something that frustrated me and I'd quit. Thank you, whoever made this one. I can't say I understood the math functions at all (I'm an artist, not a programmer). And the relationship between nodes and their inherited properties/methods either. Like it took me forever to figure out how I was able to rotate the sprite (that it inherits that property from Node2D or whatever. I just might be able to make my RPG someday. Freakin love Godot

**Comment by leowribeiro** [comment]
Guys I made my first game on Godot based on Dodge the Creeps, I called it "Squids" check it out!
https://youtu.be/QkIBNgWTg2c
I tried to put some spark in Dodge the Creeps, please tell me what you think about it!

**Reply by RobbiewOnline** [reply]
Gosh your high score increments extremely fast, I'd reduce the pace to be at most once per second, otherwise it's difficult to track and be motivated to beat such a high and difficult to grok number.

The launch graphic looks very nice, the main game play almost looks identical, except the addition of the health bar, so no instant death.

The ending has been improved.

Well done for making some tweaks to make your first game. How about changing more of the gameplay, for example different coloured enemy, some may change direction or speeds, some might blow up, some might spawn babies, etc

**Reply by nitesh098suthar** [reply]
I visited mentioned video and your game is definitely great. BTW nice background music, 
How did you design that sound?

**Comment by lowbit** [comment]
Setting players/mobs or both collision shapes a bit smaller then the sprite makes for more enjoyable experience where the player gets better satisfaction in near misses and might get less annoyed at deaths by blaming himself being bad then the game.

**Reply by Calinou** [reply]
On the other hand, making their collision shapes smaller will make squashing more difficult. Ideally, you'd make it so the player's hitbox for squashing is larger than its actual collision with mobs if you were to go that route.

**Comment by callumscreativearts** [comment]
Hello, I finished this tutorial and only encountered one problem (which isn't all too bad since it's my first time working with Godot). Everything worked until I put the ColorRect node into the scene tree. When the ColorRect node was visible, my player would be hidden, even though the ColorRect node was underneath all my main components. Other than that, my game functions as it should, but it would be nice to fix the issue with the background and not seeing the player. The mobs still appear, and the the HUD is visble, only the player is invisible (but very much still there and colliding with the mobs) when the ColorRect node enters my project. Great guide though!

**Reply by callumscreativearts** [reply]
actually, disregard my issue I have figured out the problem :P

**Reply by Hustle767** [reply]
how did you fix this?

**Comment by JSziklai** [comment]
I just completed this tutorial, and it is awesome! In the beginning it is very detailed, but after a while you are forced to take the steps on your own... kudos to the authors!
I did have one issue. My mobs worked just fine, but they didn't animate. Did I miss a step?
My solution was to add ` $AnimatedSprite2D.play() ` after ` $AnimatedSprite2D.animation = mob_types.pick_random() ` in the ` _ready() ` function of the mob.gd file.

**Reply by associateone** [reply]
If I remember correctly, the tutorial said to put it in the mob.gd _process(delta) function. I believe your solution is best. An equal solution would be to set each animation to autoplay on load in the SpriteFrames editor.

**Reply by JSziklai** [reply]
The tutorial never mentions `mob.gd`, it only says "Save the scene."
In my opinion putting the `$AnimatedSprite2D.play()` command in the `_process(delta)` function would be wrong, since each instance only needs to be started once.
Thank you for the "Autoplay on Load" option, maybe the tutorial should also mention this.

**Reply by associateone** [reply]
I did this tutorial a couple years ago and was curious, so I went back and checked. It never does say where to put it.
I think when I was done with it, I noticed the mobs weren't animating so wanting to get it over and move on to the 3d tutorial I just put it the first place I knew it would work.
Your solution is by far the best choice. Well done.

**Comment by Sirrender-krr** [comment]
> # getting_started/first_2d_game/07.finishing-up
> We have now completed all the functionality for our game. Below are some remaining steps to add a bit more "juice" to improve the game experience. Feel free to expand the gameplay with your own ide...
> 
> https://docs.godotengine.org/en/latest/getting_started/first_2d_game/07.finishing-up.html

I just finished the tutorial project, it is fun and educational. I have adjust it using knowledge that I have gained. Feel free to try.
https://nearamong.itch.io/dodge-100

**Comment by TheoMacario** [comment]
Hey Guys Idk what I did wrong but my "game over" Screen just wont apear. Also everything related to it like the deathsound also wont come up. Does someboddy know what i can do to fix it?

**Reply by Aspherioth** [reply]
It is difficult to pinpoint the cause without more information, but if you're still having issues here are a few things you can try looking into:

1. On your player script, check if you have the "hit" signal setup and you're running hit.emit().
2. On your Main scene, select the Player node and on the right side of the window go to Node > Signals and check if the hit signal has a method linked to it (just below the hit signal there should be a green icon and the name of the connected method).
3. Double-check if the game_over code is similar to the one in the tutorial.

**Comment by ohineedascreenname** [comment]
When I run the game, all of my mob sprites appear in the same direction (with their legs to the left side). I assume this has something to do with my mob script:

<img width="494" height="308" alt="Mob script" src="https://github.com/user-attachments/assets/2d5cf480-3b56-4906-a921-13c999caef12" />

<img width="514" height="748" alt="image" src="https://github.com/user-attachments/assets/8dfd7185-d4bc-4f21-ac1a-650c08a8b64e" />

**Reply by associateone** [reply]
Been awhile, but I think it has to do with the path and how the MOB is spawned.

**Reply by ohineedascreenname** [reply]
I just added to the `mob` script's `func _ready()` function this line: `$AnimatedSprite2D.rotate(randi())` and now they all enter at random rotated angles.

**Reply by associateone** [reply]
Good deal, putting it in the creation sequence is the way.

**Comment by associateone** [comment]
Here's how the tutorial does it which provides for flexibility on what you do with the MOB...

```gdscript
func _on_mob_timer_timeout():
	# Create a new instance of the Mob scene.
	var mob = mob_scene.instantiate()

	# Choose a random location on Path2D.
	var mob_spawn_location = $MobPath/MobSpawnLocation
	mob_spawn_location.progress_ratio = randf()

	# Set the mob's position to the random location.
	mob.position = mob_spawn_location.position

	# Set the mob's direction perpendicular to the path direction.
	var direction = mob_spawn_location.rotation + PI / 2

	# Add some randomness to the direction.
	direction += randf_range(-PI / 4, PI / 4)
	mob.rotation = direction

	# Choose the velocity for the mob.
	var velocity = Vector2(randf_range(150.0, 250.0), 0.0)
	mob.linear_velocity = velocity.rotated(direction)

	# Spawn the mob by adding it to the Main scene.
	add_child(mob)
```

**Comment by lil-sassy** [comment]
I put together a little follow-along video. I figured this might be helpful for some people to see someone else go through this small project after finishing it and how it should look.

https://www.youtube.com/watch?v=hljPmVJEOkY

**Comment by fawazadydrm** [comment]
i just get problems because when i start a game, the text didnt remove and when i start a new game again it didnt respond anything, and it's just started even i didn't press the start button

**Comment by poorforever** [comment]
So before reading the section about shortcuts, I was already trying to make the Enter key, and achieved it using two lines of script in main.gd. After defining the input "start_game" mapped to the Enter key or controller key, you can use this:

```gdscript
func _process(_delta: float) -> void:
	if Input.is_action_pressed("start_game"):
		$HUD/StartButton.emit_signal("pressed")
```

Maybe it'll be useful in some use cases.

**Comment by cmay116** [comment]
If anyone's Enter key shortcut isn't working, make sure you're connected to the pressed() signal. I was using the button_down() signal by mistake, so my button click from the cursor would work, but the InputEventAction shortcut was not working.

Fun tutorial!

**Comment by oOWhattahWOo** [comment]
Hi guys, I ran into an issue when importing BGM that caused my `main.tscn` to suddenly bloat to 1.84 MiB. Godot shows this warning:
```
The text-based scene at path "res://main.tscn" is large on disk (1.84 MiB), likely because it has embedded binary data.
This slows down scene saving and loading.
Consider saving its binary subresource(s) to a binary .res file or saving the scene as a binary .scn file.
```
After digging into it (and checking with AI), I found the culprit was this step from the tutorial: **“If you want the music to loop seamlessly, click on the Stream file arrow, select Make Unique”**

It looks like clicking `Make Unique` copies the external audio resource into the scene as a SubResource. If I open `main.tscn` in a text editor, I can see a massive block starting with: `[sub_resource type="OggPacketSequence" id="OggPacketSequence_1utue"]`

The whole OGG data is embedded there, which explains the file size jump.

So my question is:

👉 What’s the best way to enable BGM looping without having the audio embedded into the `.tscn` file?
👉 Is there a recommended workflow for this?
👉 Any good docs I should read to better understand how Godot handles imported audio resources?

Thanks in advance!

**Reply by oOWhattahWOo** [reply]
I think I found a clean fix for the “huge main.tscn” issue when enabling BGM loop — and it keeps the scene file small (a few KB) while the music still loops correctly.

If there’s a better workflow or any solid docs that explain this more systematically, **please share!**

Here’s exactly what I did:

1️⃣ Remove the embedded stream

Click the `Music` node

In the `Inspector`, find the `Stream` property

Click the small revert/clear (undo) icon next to Stream

Save the scene (Ctrl + S)

At this point the BGM is removed, and `main.tscn` drops back down to just a few KB.

2️⃣ Enable Loop from the Import settings (not Make Unique)

In the bottom-left `FileSystem` panel, select:
`House In a Forest Loop.ogg`

At the top panel, switch to the `Import` tab

Check the `Loop` option

Click `Reimport`

3️⃣ Add the music back and save

Reassign the same `.ogg` file to the `Music` node’s `Stream` and save `main.tscn` again

Now:

`main.tscn` stays only a few KB

The BGM loops correctly

**Reply by DavidDubinin** [reply]
hi man, thanks for the workaround!

**Reply by WojtekWZ** [reply]
The response is in the warning - you can save the scene as a main.scn file using scene -> save scene as -> choose .scn extension. Then you can delete the main.tscn file. One thing to note, switching to scn means the file is binary and not as easy to work with version control.

**Reply by phamtrongngh** [reply]
thank you

**Comment by SirPeteThe1st** [comment]
I too am lost in several ways. 
1) I also get the greyscreen error. I followed the lessons very carefully, compared my written code to the source, and there is no error log other than the greyscreen when running. 

2) Referencing the following instructions:
"In the HUD scene, select the StartButton and find its Shortcut property in the Inspector. Create a new Shortcut resource by clicking within the box, open the Events array and add a new array element to it by clicking on Array[InputEvent] (size 0)." 

I cannot locate anything in the illustration provided. I am running Godot engine on macOS. 

Please advise.

**Reply by WojtekWZ** [reply]
In the hud scene, you select the StartButton from the scene tree on the left. Then on the right side, in the Inspector, you should see button properties. The list of properties should be divided in sections : Button, BaseButton, Control... Shortcut is in the BaseButton. If you can't find it, there's also property search dialog at the top of Inspector, so just type `shortcut` in it.

**Reply by WojtekWZ** [reply]
And re: 1st point - are you sure you're running the correct scene? Did you check if there's anything in the logs?

**Comment by admin** [comment]
*Originally posted by [**oxeydane**](https://github.com/oxeydane) on 2026-06-08 11:40:52 UTC*

> If you're stuck while thinking what to make, make a tic-tac-toe game now

**Reply by admin** [reply]
*Originally posted by [**NightHawk5631**](https://github.com/NightHawk5631) on 2026-06-11 14:50:10 UTC*

> Would that teach me anything?

**Reply by admin** [reply]
*Originally posted by [**xespera**](https://github.com/xespera) on 2026-06-23 22:01:50 UTC*

> Depends on where you are on your journey. Tic-Tac-Toe would create a few challenges to figure out.
>
> Mouse Input for picking which square, figuring out "When is the game over" (Both all spots filled And detecting a win, including if both happen at the same time), and "How to get the computer to figure out best moves to make"
